### PR TITLE
feat(infra/cluster/sponsored): add node pools

### PR DESCRIPTION
**Pull request by smerle33**
as per https://github.com/jenkins-infra/helpdesk/issues/3923
and following #715 

this PR create 3 nodes pools:
- application one in arm64
- agents in x86-64
- agents in arm64

### CODE REVIEW

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 1]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+resource "azurerm_kubernetes_cluster_node_pool" "linux_x86_64_n4_agents_1_sponsorship" {
+  provider              = azurerm.jenkins-sponsorship
+  name                  = "lx86n3agt1"
+  vm_size               = "Standard_D8ads_v5" # Standard_D8ads_v5 	8 	32 	300 	16 	38000 / 500 	12800/200 	20000/600 	4 	12500
+  os_disk_type          = "Ephemeral"
+  os_disk_size_gb       = 300 # Ref. Cache storage size at https://learn.microsoft.com/en-us/azure/virtual-machines/dasv5-dadsv5-series (depends on the instance size)
+  orchestrator_version  = local.kubernetes_versions["infracijenkinsio_agents_1"]
+  kubernetes_cluster_id = azurerm_kubernetes_cluster.infracijenkinsio_agents_1.id
+  enable_auto_scaling   = true
+  min_count             = 1
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:104]:
```suggestion
  min_count             = 0
```

This node pool is not using arm64 neither spot instance

Comment (smerle33) [reply infraci.jenkins.io-kubernetes-sponsored-agents.tf:104]: just choose 1 as minimal to not have to wait for the spawning when needed ...

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 2]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+resource "azurerm_kubernetes_cluster_node_pool" "linux_arm64_n4_1_sponsorship" {
+  provider              = azurerm.jenkins-sponsorship
+  name                  = "la64n41"
+  vm_size               = "Standard_D8pds_v5" # Standard_D8pds_v5 	8 	32 	300 	16 	38000/500 	12800/290 	20000/1200 	4 	12500
+  os_disk_type          = "Ephemeral"
+  os_disk_size_gb       = 300 # Ref. Cache storage size at https://learn.microsoft.com/en-us/azure/virtual-machines/dv3-dsv3-series#dsv3-series (depends on the instance size)
+  orchestrator_version  = local.kubernetes_versions["infracijenkinsio_agents_1"]
+  kubernetes_cluster_id = azurerm_kubernetes_cluster.infracijenkinsio_agents_1.id
+  enable_auto_scaling   = true
+  min_count             = 1
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:134]:
```suggestion
  min_count             = 0
```

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 3]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+  kubernetes_cluster_id = azurerm_kubernetes_cluster.infracijenkinsio_agents_1.id
+  enable_auto_scaling   = true
+  min_count             = 0
+  max_count             = 20
+  zones                 = local.infracijenkinsio_agents_1_compute_zones
+  vnet_subnet_id        = data.azurerm_subnet.infraci_jenkins_io_kubernetes_agent_sponsorship.id
+
+  node_labels = {
+    "jenkins" = "infra.ci.jenkins.io"
+    "role"    = "jenkins-agents-x86-64"
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:111]:
```suggestion
    "role"    = "jenkins-agents"
```

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 4]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+  kubernetes_cluster_id = azurerm_kubernetes_cluster.infracijenkinsio_agents_1.id
+  enable_auto_scaling   = true
+  min_count             = 0
+  max_count             = 20
+  zones                 = local.infracijenkinsio_agents_1_compute_zones # need to be on zone 1 for arm availability
+  vnet_subnet_id        = data.azurerm_subnet.infraci_jenkins_io_kubernetes_agent_sponsorship.id
+
+  node_labels = {
+    "jenkins" = "infra.ci.jenkins.io"
+    "role"    = "jenkins-agents-arm64"
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:141]:
```suggestion
    "role"    = "jenkins-agents"
```

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 5]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+  ]
+
+  lifecycle {
+    ignore_changes = [node_count]
+  }
+
+  tags = local.default_tags
+}
+
+# Node pool to host infra.ci.jenkins.io agents for BOM builds
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:124]:
```suggestion
# Node pool to host infra.ci.jenkins.io arm64 agents
```

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 6]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+  ]
+
+  lifecycle {
+    ignore_changes = [node_count]
+  }
+
+  tags = local.default_tags
+}
+
+# Node pool to host infra.ci.jenkins.io agents for usual builds
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:94]:
```suggestion
# Node pool to host infra.ci.jenkins.io x86_64 agents
```

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 7]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+  node_taints = [
+    "infra.ci.jenkins.io/applications=true:NoSchedule",
+  ]
+
+  lifecycle {
+    ignore_changes = [node_count]
+  }
+
+  tags = local.default_tags
+}
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:92]:
```suggestion
```

pair review: the only expected "application" is the Datadog's cluster-agent (2 pods). We should reuse the system pool to run it as it's not an heavy consumer: confirmed by checking the load on the ci.jenkins.io-agents-1 cluster with the same kind of node pool.

@smerle33 to comment the issue to cherry-pick this choice and arguments

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 8]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+
+# Node pool to host infra.ci.jenkins.io agents for usual builds
+resource "azurerm_kubernetes_cluster_node_pool" "linux_x86_64_n4_agents_1_sponsorship" {
+  provider              = azurerm.jenkins-sponsorship
+  name                  = "lx86n3agt1"
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:97]:
```suggestion
  name                  = "lx86n14agt1"
```

Comment (dduportal) [reply infraci.jenkins.io-kubernetes-sponsored-agents.tf:97]: - Need to add resource limit reference for `jnlp-linux`
- Update comment below if needed

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 9]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+    ignore_changes = [node_count]
+  }
+
+  tags = local.default_tags
+}
+
+# Node pool to host infra.ci.jenkins.io agents for BOM builds
+resource "azurerm_kubernetes_cluster_node_pool" "linux_arm64_n4_1_sponsorship" {
+  provider              = azurerm.jenkins-sponsorship
+  name                  = "la64n41"
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:127]:
```suggestion
  name                  = "la64n14agt1"
```

- Need to add resource limit reference for `jnlp-linux`
- Update comment below if needed

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 10]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+
+  lifecycle {
+    ignore_changes = [node_count]
+  }
+
+  tags = local.default_tags
+}
+
+# Node pool to host infra.ci.jenkins.io agents for BOM builds
+resource "azurerm_kubernetes_cluster_node_pool" "linux_arm64_n4_1_sponsorship" {
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:125]:
```suggestion
resource "azurerm_kubernetes_cluster_node_pool" "linux_arm64_agents_1_sponsorship" {
```

File: infraci.jenkins.io-kubernetes-sponsored-agents.tf [hunk 11]
```diff
@@ -59,3 +59,94 @@ resource "azurerm_kubernetes_cluster" "infracijenkinsio_agents_1" {
...
+
+  lifecycle {
+    ignore_changes = [node_count]
+  }
+
+  tags = local.default_tags
+}
+
+# Node pool to host infra.ci.jenkins.io agents for usual builds
+resource "azurerm_kubernetes_cluster_node_pool" "linux_x86_64_n4_agents_1_sponsorship" {
```
Comment (dduportal) [inline infraci.jenkins.io-kubernetes-sponsored-agents.tf:95]:
```suggestion
resource "azurerm_kubernetes_cluster_node_pool" "linux_x86_64_agents_1_sponsorship" {
```

**Review by dduportal — APPROVED** [review]
`Plan: 2 to add, 0 to change, 0 to destroy.`

=> LGTM!

@smerle33 you can merge whenevr you want!

**Pull request merged**